### DISCUSSION TITLE
Trimming LLM responses

### DISCUSSION THREAD

**Discussion by ProjectMoon** [karthink/gptel#336] [Q&A] [answered]
I'm working on making a prompt/settings that I will bind to keys for completing a prose paragraph. The problem is that often, the models write their responses a few lines below my cursor, or with spaces in front. 

Is there a way using the post stream hook to trim the whitespace and have the LLM response begin from the cursor?

I know it is probably pretty easily possible, I just don't know enough elisp to do this properly.

**Comment by karthink** [comment, marked as answer]
You can do this from the post response hook:

```elisp
(defun gptel-trim-leading-whitespace (beg _)
  "Delete leading whitespace before replies."
  (unless gptel-mode
    (save-excursion
      (goto-char beg)
      (delete-region
       (point) (progn (skip-syntax-backward " ")
                      (point))))))

(add-hook 'gptel-post-response-functions #'gptel-trim-leading-whitespace)
```

Should the default behavior of gptel in regular (i.e. non chat) be to not inject any whitespace?

**Reply by ProjectMoon** [reply]
Thanks for the snippet. I did something similar, but with just deleting backwards twice. This is much more elegant.

I think it might be good for this to be the default in prose contexts. Possibly toggleable with a variable or via use of a separate command/transient option?

**Reply by karthink** [reply]
I think perhaps a single setting that controls "chat-like" behavior in regular buffers might be useful.  This can include the formatting (like the newlines before the response), as well as differentiating between regular and response text.  The latter is currently controlled by `gptel-track-response`.

**Reply by ProjectMoon** [reply]
Would also be a good foundation for your experiments with auto complete, perhaps.

**Reply by ProjectMoon** [reply]
A setting is probably the easiest solution. Allows easy toggling as well as mode hook integration.
